### PR TITLE
Private data tutorial edits

### PR DESCRIPTION
**Pull request by nikhil550**
Signed-off-by: NIKHIL E GUPTA <ngupta@symbridge.com>

#### Type of change

- Documentation update

#### Description

Use the deployCC script to deploy the private data smart contract. The tutorial is already quite long, so this change will make life easier for users. 

Also made some minor edits

### CODE REVIEW

**Comment by denali49** [comment]
While I agree that the scripted approach makes life easier, it may be defeating the purpose of a 'tutorial' even if it is a bit long.  Most maintainers work with Fabric every day and the mundane bits can be dispensed with, but for the new user exposing them to the 'how' is beneficial to learning Fabric.  At the very least, we should give the user the option of taking the easy way out with the script, or going through the long tutorial.  My two cents.

**Comment by nikhil550** [comment]
> While I agree that the scripted approach makes life easier, it may be defeating the purpose of a 'tutorial' even if it is a bit long. Most maintainers work with Fabric every day and the mundane bits can be dispensed with, but for the new user exposing them to the 'how' is beneficial to learning Fabric. At the very least, we should give the user the option of taking the easy way out with the script, or going through the long tutorial. My two cents.

I am sympathetic to this argument. That is why I left the commands to deploy the collection configuration along with the script, and linked to the chaincode lifecycle topic. There is a separate tutorial that helps users use the cc lifecycle. Note that the other tutorials in this series use the deployment script as well.

My worry is that the long tutorial will discourage users, and that they will not get to the private data parts of the tutorial. I also think that keeping the documentation modular will help users. They can use the private data tutorial for private data, couchdb for couchDB, and the deployment tutorial for lifecycle. I would be interested in what others think on this.

File: docs/source/private_data_tutorial.rst [hunk 1]
```diff
@@ -434,202 +418,45 @@ use indexes with private data.
...
-The command will return the same package identifier as the install command.
-You should see output similar to the following:
-
-.. code:: bash
+   ./network.sh deployCC -ccn private -ccep "OR('Org1MSP.peer','Org2MSP.peer')" -cccg ../asset-transfer-private-data/chaincode-go/collections_config.json
 
-    Installed chaincodes on peer:
-    Package ID: privatev1:c10ad5913bd3166c1bcc294bb378ecef5f815b516b717f98d8a68ff2f6f4eee9, Label: privatev1
+Note that we need to pass the path to the private data collection definition file
+to the command. As part of the deploying the chaincode to the channel, both organizations
```
Comment (denyeart) [inline docs/source/private_data_tutorial.rst]:
```suggestion
to the command. As part of deploying the chaincode to the channel, both organizations
```

File: docs/source/private_data_tutorial.rst [hunk 2]
```diff
@@ -434,202 +418,45 @@ use indexes with private data.
...
+on the channel need to agree deploy the private data collection as part
```
Comment (denyeart) [inline docs/source/private_data_tutorial.rst]:
```suggestion
on the channel must pass identical private data collection definitions as part 
```

File: docs/source/private_data_tutorial.rst [hunk 3]
```diff
@@ -1083,6 +881,79 @@ If you go back and look at the block height in your logs, you will see that the
...
 because when the asset transfer occurred, Org1 did not meet the ``read`` policy of the ``Org2MSPPrivateCollection``.
 
+
+.. _pd-purge:
+
+Purge Private Data
+------------------
+
+For use cases where private data only needs to be on the ledger until it can be
+replicated into an off-chain database, it is possible to "purge" the data after
```
Comment (denyeart) [inline docs/source/private_data_tutorial.rst]: replication to an off-chain database is one use case, but there are other use cases. So perhaps say something more general like, e.g.
"In some cases private data only needs to be persisted for a short time, for example while a multi-party business process is in progress, or until the data is replicated into an off-chain database"...

File: docs/source/private_data_tutorial.rst [hunk 4]
```diff
@@ -1083,6 +881,79 @@ If you go back and look at the block height in your logs, you will see that the
...
+a certain set number of blocks, leaving behind only hash of the data that serves
```
Comment (denyeart) [inline docs/source/private_data_tutorial.rst]:
```suggestion
a certain set number of blocks, leaving behind only a hash of the data that serves
```

File: docs/source/private_data_tutorial.rst [hunk 5]
```diff
@@ -1083,6 +881,79 @@ If you go back and look at the block height in your logs, you will see that the
...
+as immutable evidence of the transaction.
+
+There may be private data including personal or confidential
+information, such as the ``appraisedValue`` data in our example, that the transacting
+parties don't want disclosed to other organizations on the channel. Thus, it
```
Comment (denyeart) [inline docs/source/private_data_tutorial.rst]: Private data in a collection is never disclosed to non-member peers, so I wouldn't say that is the reason to purge it.
Focus on the limited lifespan aspect.  The appraisedValue scenario is good, as you may want some time limit on the agreement (of course, number of blocks is a very rough way to specify time).

File: docs/source/private_data_tutorial.rst [hunk 6]
```diff
@@ -1083,6 +881,79 @@ If you go back and look at the block height in your logs, you will see that the
...
+has a limited lifespan, and can be purged after existing unchanged on the
+blockchain for a designated number of blocks using the ``blockToLive`` property
+in the collection definition.
+
+The ``Org2MSPPrivateCollection`` definition has a ``blockToLive``
+property value of ``3``, meaning this data will live on the side database for
+three blocks and then after that it will get purged. If we create additional
+blocks on the channel, the appraisedValue agreed to by Org2 will eventually
+get purged. We can create 3 new blocks to demostrate:
```
Comment (denyeart) [inline docs/source/private_data_tutorial.rst]:
```suggestion
get purged. We can create 3 new blocks to demonstrate:
```

**Review by denyeart — COMMENTED** [review]
I like how you handle the lifecycle Nik... agreed that the detailed steps don't need to be repeated in every tutorial, and I like how you still discussed the important aspect around collection definition during approval and commit.

**Comment by denyeart** [comment]
@nikhil550 @denali49 I think there are some aspects mentioned in the [fabric-samples readme scenario description ](https://github.com/hyperledger/fabric-samples/tree/master/asset-transfer-private-data/chaincode-go#private-data-asset-transfer-scenario)that are not yet covered in the tutorial. Could you verify that all the important information from the readme is included in the tutorial, and then we can replace the readme content with a link to the doc tutorial.

**Comment by pamandrejko** [comment]
We discussed this in our Community call this morning. 

One argument for keeping the lifecycle steps in this tutorial is that you must provide the private data collection during that process. A newbie could overlook that part of the deployment process so it seems worthwhile including the steps in this tutorial so that it is very clear. This is the one case where that part of the lifecycle process is actually required.

File: docs/source/private_data_tutorial.rst [hunk 7]
```diff
@@ -37,23 +37,24 @@ You should have completed the task :doc:`install`.
...
 
 .. _pd-build-json:
 
 Build a collection definition JSON file
 ---------------------------------------
 
-The first step in privatizing data on a channel is to build a collection
-definition which defines access to the private data.
+Before a set of organizations can transact using private data on a channel, all organizations
+on channel need to build a collection definition which defines access to the private data.
```
[the remark this answers is not in the document]
Comment (nikhil550) [reply docs/source/private_data_tutorial.rst]: Rephrasing: Before a set of organizations can transact using private data, all organizations
on channel need to build a collection file that defines the private
data collections associated with each chaincode. Data that is stored in a private
data collection is only distributed to the peers of certain organizations instead
of all members of the channel. The collection definition file describes all of the
private data collections that organizations can read and write to from a chaincode.

further down:The same collection definition file needs to be deployed by all organizations that
use the chaincode, even if the organization does not belong to any collections. In
addition to the collections that are explicitly defined in a collection file,
each organization has access to an implicit collection on their peers that can only
be read by their organization. For an example that uses implicit data collections,
see the :doc:`secured_asset_transfer/secured_private_asset_transfer_tutorial`.

File: docs/source/private_data_tutorial.rst (hunk 1 above)
Comment (nikhil550) [reply docs/source/private_data_tutorial.rst]: done.

File: docs/source/private_data_tutorial.rst (hunk 2 above)
Comment (nikhil550) [reply docs/source/private_data_tutorial.rst]: done

File: docs/source/private_data_tutorial.rst (hunk 3 above)
Comment (nikhil550) [reply docs/source/private_data_tutorial.rst]: For use cases where private data only needs to be persisted for a short period of time,
it is possible to "purge" the data after a certain set number of blocks, leaving
behind only hash of the data that serves as immutable evidence of the transaction.
An organization could decide to purge private data if the data contained sensitive
information that was used by another transaction, but is not longer needed, or
if the data is being replicated into an off-chain database.

File: docs/source/private_data_tutorial.rst (hunk 5 above)
Comment (nikhil550) [reply docs/source/private_data_tutorial.rst]: The ``appraisedValue`` data in our example contains a private agreement that
the organization may want to expire after a certain period of time. Thus, it
has a limited lifespan, and can be purged after existing unchanged on the
blockchain for a designated number of blocks using the ``blockToLive`` property
in the collection definition.

File: docs/source/private_data_tutorial.rst (hunk 6 above)
Comment (nikhil550) [reply docs/source/private_data_tutorial.rst]: done

File: docs/source/private_data_tutorial.rst (hunk 4 above)
Comment (nikhil550) [reply docs/source/private_data_tutorial.rst]: done.

**Comment by nikhil550** [comment]
I can do this in a follow on PR sometime this week.

> @nikhil550 @denali49 I think there are some aspects mentioned in the [fabric-samples readme scenario description ](https://github.com/hyperledger/fabric-samples/tree/master/asset-transfer-private-data/chaincode-go#private-data-asset-transfer-scenario)that are not yet covered in the tutorial. Could you verify that all the important information from the readme is included in the tutorial, and then we can replace the readme content with a link to the doc tutorial.

Most of the information in the readme is in the tutorial. However, some of the information about which members can read and write to each collection is not in the tutorial. Most importantly, the use of the `GetPrivateDataHash()` to verify private data is not in the tutorial and should be added.

**Review by denyeart — APPROVED** [review]

**Comment by denyeart** [comment]
> We discussed this in our Community call this morning.
> 
> One argument for keeping the lifecycle steps in this tutorial is that you must provide the private data collection during that process. A newbie could overlook that part of the deployment process so it seems worthwhile including the steps in this tutorial so that it is very clear. This is the one case where that part of the lifecycle process is actually required.

@pamandrejko Nik's update includes the approve and commit commands that reference the collection definition file. This is the important aspect for people to understand.  I think highlighting only these commands is the best way to highlight that, rather than it getting buried in the full set of lifecycle commands.  The lifecycle tutorial shows the full set of lifecycle commands. This private data tutorial shows the delta difference for private data chaincodes.

**Pull request merged**